neps/nep-0481.md

```diff
@@ -0,0 +1,145 @@
+- Submodule storage: Submodule bytecode has to be stored such that it is writable only via actions `DeploySubmodule` and `RemoveSubmodule`.
+- Removing submodules: Support for the action `RemoveSubmodule` has to be added.
+- Error handling:
+  - Not all errors related to submodules that may occur inside the host are handled properly (some errors are `unwrap()`’ed in the PoC).
+  - Traps occuring during the execution of submodule bytecode need to be handled.
+
+## Alternatives
+
+Possible alternatives for the synchronous execution of wasm bytecode are:
+
+### An `eval` host functions
+
+An `eval` host function that executes bytecode which it received as input parameter.
+
+Pros:
+
+- Might allow the execution of bytecode without previous deployment, which could reduce the complexity of some use cases.
+
+Cons:
+
+- Coroutines are meant to run concurrently and be resumable, which are features required for supporting the concurrent execution of submodules. Supporting this functionality with an `eval` function might be more involved.
+- Handling traps in submodules is likely to be more challenging when the submodule is executed in the same virtual machine as the parent.
+- In some programming languages similar `eval` functions have a notoriously bad reputation, see for instance [Python](https://stackoverflow.com/questions/1832940/why-is-using-eval-a-bad-practice/1832957#1832957) and [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#never_use_eval!).
+
+### Account extensions
+
+_Note that the discussion regarding the scope and specification of account extensions is still ongoing. As of 2023-05-10, a NEP for account extensions does not yet exist._
+
+The synchronous execution of wasm bytecode should, ultimately, be possible with account extensions. In that case, setting appropriate submodule permissions should serve the use cases that would be enabled by this proposal. However, account extensions are much broader in scope and accordingly their implementation is expected to be significantly more complex. Instead of being an alternative, synchronous wasm submodules might rather be a building block for account extensions.
+
+## Drawbacks
+
+### Passing data by exchanging bytes
+
+Limiting the interface between a parent and its submodules to passing bytes introduces overhead in cases that require interaction. For instance, instead of directly persisting data in storage a submodule has to serialize data and send it back to the parent. The parent then needs to deserialize the data, write it to storage and resume the submodule. Besides requiring extra logic this pattern also increases the number of host function calls.
+
+That overhead could be reduced by extending the interface, for instance by making more host functions available to submodules. The trade-off with giving submodules direct access to more host functions is complexity in permissions.
```

Comment:
This seems like something that could be implemented by introducing a mechanism for the "start submodule" to specify which host functions are available to the callee.